### DISCUSSION TITLE
Audio is choppy/broken up

### DISCUSSION THREAD

**Discussion by Colin1964** [pndurette/gTTS#277] [Q&A] [answered]
Using PI OS (BUSTER) on RPI3 and Python 3.7.3. When using either the CLI or module version of gTTS and passing the output to a MP3 file the audio that is produced is "choppy" with either words missing or long pauses. The problem seems to occur for all but very short pieces of text. E.g. with a a text phrase such as "The sun is shining" the cli and module versions will output the audio without error. If the phrase is increased to say "The sun is shining today but it will rain tomorrow" the audio produced is something like "[pause] is shining today [pause] will [pause] tomorrow". Any ideas what may be causing this and how I might fix it?

**Comment by pndurette** [comment, marked as answer]
@Colin1964 Hi!

Hmm, so you're saying that `The sun is shining today but it will rain tomorrow` is choppy, but not `The sun is shining today`? I just tried here (Python 3.8, macOS) and can't replicate.

Let me look at the debug. Can do it again with `gtts-cli` and add the `--debug` flag and send the output?

Thanks!

**Reply by Colin1964** [reply]
@pndurette many thanks for the speedy response. Debug output is below but I have discovered the problem is with playback on the Pi and not gTTS - sorry! I transferred over the choppytest.mp3 file over to my PC from my Pi so I could send you a google drive link so that you could hear the output I was getting from the Pi. I played the mp3 on my PC and it worked flawlessly - no gaps or pauses (hear it for yourself here: https://drive.google.com/file/d/19wdy2YnCJ1ZyeJ_OyjUoGvYCGuYkPkU-/view?usp=sharing) So it looks like the vagaries of getting sound to work with Linux on RPi is my issue.  I'll work through that now but it's clear gTTS is not at fault here - thanks again and great work BTW!

pi@raspberrypi:~ $ gtts-cli --debug "The sun is shining today but rain expected tomorrow" -o choppytest.mp3
gtts.lang - DEBUG - langs: {'af': 'Afrikaans', 'ar': 'Arabic', 'bn': 'Bengali', 'bs': 'Bosnian', 'ca': 'Catalan', 'cs': 'Czech', 'cy': 'Welsh', 'da': 'Danish', 'de': 'German', 'el': 'Greek', 'en': 'English', 'eo': 'Esperanto', 'es': 'Spanish', 'et': 'Estonian', 'fi': 'Finnish', 'fr': 'French', 'gu': 'Gujarati', 'hi': 'Hindi', 'hr': 'Croatian', 'hu': 'Hungarian', 'hy': 'Armenian', 'id': 'Indonesian', 'is': 'Icelandic', 'it': 'Italian', 'ja': 'Japanese', 'jw': 'Javanese', 'km': 'Khmer', 'kn': 'Kannada', 'ko': 'Korean', 'la': 'Latin', 'lv': 'Latvian', 'mk': 'Macedonian', 'ml': 'Malayalam', 'mr': 'Marathi', 'my': 'Myanmar (Burmese)', 'ne': 'Nepali', 'nl': 'Dutch', 'no': 'Norwegian', 'pl': 'Polish', 'pt': 'Portuguese', 'ro': 'Romanian', 'ru': 'Russian', 'si': 'Sinhala', 'sk': 'Slovak', 'sq': 'Albanian', 'sr': 'Serbian', 'su': 'Sundanese', 'sv': 'Swedish', 'sw': 'Swahili', 'ta': 'Tamil', 'te': 'Telugu', 'th': 'Thai', 'tl': 'Filipino', 'tr': 'Turkish', 'uk': 'Ukrainian', 'ur': 'Urdu', 'vi': 'Vietnamese', 'zh-CN': 'Chinese', 'zh-TW': 'Chinese (Mandarin/Taiwan)', 'zh': 'Chinese (Mandarin)'}
gtts.tts - DEBUG - text: The sun is shining today but rain expected tomorrow
gtts.tts - DEBUG - tld: com
gtts.tts - DEBUG - lang: en
gtts.tts - DEBUG - slow: False
gtts.tts - DEBUG - lang_check: False
gtts.tts - DEBUG - pre_processor_funcs: [<function tone_marks at 0x76667858>, <function end_of_line at 0x76667a98>, <function abbreviations at 0x76667bb8>, <function word_sub at 0x76667fa8>]
gtts.tts - DEBUG - tokenizer_func: <bound method Tokenizer.run of re.compile('(?<=\\?).|(?<=!).|(?<=？).|(?<=！).|(?<!\\.[a-z])\\. |(?<!\\.[a-z]), |(?<!\\d):|…|;|\\)|\\\n|¿|‥|\\[|\\(|，|—|،|\\]|：|¡|、|。', re.IGNORECASE) from: [<function tone_marks at 0x765f9078>, <function period_comma at 0x765f90c0>, <function colon at 0x765f9108>, <function other_punctuation at 0x765f9150>]>
gtts.tts - DEBUG - pre-processing: <function tone_marks at 0x76667858>
gtts.tts - DEBUG - pre-processing: <function end_of_line at 0x76667a98>
gtts.tts - DEBUG - pre-processing: <function abbreviations at 0x76667bb8>
gtts.tts - DEBUG - pre-processing: <function word_sub at 0x76667fa8>
gtts.tts - DEBUG - text_parts: ['The sun is shining today but rain expected tomorrow']
gtts.tts - DEBUG - text_parts: 1
gtts.tts - DEBUG - data-0: f.req=%5B%5B%5B%22jQ1olc%22%2C%22%5B%5C%22The%20sun%20is%20shining%20today%20but%20rain%20expected%20tomorrow%5C%22%2C%5C%22en%5C%22%2Cnull%2C%5C%22null%5C%22%5D%22%2Cnull%2C%22generic%22%5D%5D%5D&
gtts.tts - DEBUG - headers-0: {'Referer': 'http://translate.google.com/', 'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36', 'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8', 'Content-Length': '197'}
gtts.tts - DEBUG - url-0: https://translate.google.com/_/TranslateWebserverUi/data/batchexecute
gtts.tts - DEBUG - status-0: 200
gtts.tts - DEBUG - part-0 written to <_io.BufferedWriter name='choppytest.mp3'>
pi@raspberrypi:~ $ cvlc choppytest.mp3
VLC media player 3.0.11 Vetinari (revision 3.0.11-0-gdc0c5ced72)
[0063dec8] dummy interface: using the dummy interface module...

**Reply by Colin1964** [reply]
@pndurette Issue resolved!! Found the cure for my stuttering RPi here: [https://baheyeldin.com/linux/how-fix-audio-skips-and-stuttering-raspberry-pi.html](url). The upgrade to firmware did the trick and my Digital Assistant, Gladys now has a voice with soft female soothing tones as opposed to the robot tones of Marvin that was the best that Espeak had to offer. V. happy :)
